proxy/mitmproxy.py

```diff
@@ -0,0 +1,125 @@
+import os
+import signal
+import subprocess
+import sys
+
+dirname = os.path.dirname(__file__)
+
+
+class MITMProxyBase(object):
+    def __init__(self, path, version, mode):
+        self.path = path
+        self.mode = mode
+        self.version = version
+        self.mitm_home = os.path.join(os.path.expanduser("~"), ".mitmproxy")
+        self.cert = os.path.join(self.mitm_home, "mitmproxy-ca-cert.cer")
+        self.scripts = os.path.join(dirname, "scripts")
+        self.binary_path = os.path.join(dirname, "utils", "mitm%s" % self.version)
+
+    @property
+    def binary(self):
+        if "linux" in sys.platform:
+            name = "mitmdump-linux"
+        elif "darwin" in sys.platform:
+            name = "mitmdump-osx"
+        elif "win" in sys.platform:
+            name = "mitmdump-win.exe"
+        return os.path.join(self.binary_path, name)
+
+    def start(self):
+        print("Starting MitmProxy")
+        print(" ".join(self.command()))
+
+        logfile = open(os.path.join(dirname, "mitmproxy.log"), "w")
+
+        self.process = subprocess.Popen(self.command(), stdout=logfile, stderr=logfile)
+
+        return self.process
+
+    def stop(self):
+        self.process.send_signal(signal.SIGINT)
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, *args):
+        try:
+            self.process.wait()
+        finally:
+            self.stop()
+
+
+class MITMProxy202(MITMProxyBase):
+    def __init__(self, path, mode="record"):
+        MITMProxyBase.__init__(self, path=path, version="2.0.2", mode=mode)
+
+    def command(self):
+        if self.mode is "record":
+            command = [
+                self.binary,
+                "--wfile",
+                self.path,
+                "--script",
+                " ".join([os.path.join(self.scripts, "inject-deterministic.py")]),
+            ]
+        elif self.mode is "replay":
+            command = [
+                self.binary,
+                "--replay-kill-extra",
+                "--script",
+                " ".join(
+                    [
+                        os.path.join(self.scripts, "alternate-server-replay-2.0.2.py"),
+                        self.path,
+                    ]
+                ),
+            ]
+        else:
+            raise Exception("Unknown proxy mode!")
```

Comment:
Nice!

Reply:
The exception should mention the proxy mode, to help with debugging failures.